dbos-rules.ts

```diff
@@ -262,32 +255,198 @@ const awaitsOnNotAllowedType: DetChecker = (node, _fn, _isLocal) => {
+  */
+
+  /* If the node's value is undefined, it hasn't been explored yet.
+  If it's false, it's not LR. If it's true, it's LR, or currently being
+  computed (which can indicate the existance of a reference cycle).
+
+  Also, it's worthy of noting that I'm not doing this result caching
+  for the sake of efficiency, it's so that reference cycles won't result
+  in infinite recursion. */
+  let nodeLRResults: Map<Node, boolean> = new Map();
+
+  function isLRWithoutResultCache(node: Node): boolean {
+    if (Node.isStringLiteral(node)) {
+      return true;
+    }
+    /* i.e. if it's a format string (like `${foo} ${bar} ${baz}`).
+    I am not supporting tagged template expressions, since they involve
+    a function call. */
+    else if (Node.isTemplateExpression(node)) {
+      return node.getTemplateSpans().every((span) => {
+        // The first child is the contained value, and the second child is the end of the format specifier
+        if (span.getChildCount() !== 2) panic("Unexpected child count for a template expression span!");
+        return isLR(span.getChildAtIndex(0));
+      });
+    }
+    else if (Node.isIdentifier(node)) {
+      for (const rvalueAssigned of getRValuesAssignedToIdentifier(fnDecl, node)) {
+        if (rvalueAssigned === "NotRValueButFnParam" || !isLR(rvalueAssigned)) return false;
+      }
+
+      return true;
+    }
+    else if (Node.isBinaryExpression(node)) {
+      return isLR(node.getLeft()) && isLR(node.getRight());
+    }
+    else {
+      return false;
+    }
+  }
+
+  // TODO: if it was not LR, then store the node that failed the check, in order to get the right line number
+  function isLR(node: Node): boolean {
+    const maybeResult = nodeLRResults.get(node);
+
+    if (maybeResult !== undefined) {
+      return maybeResult;
+    }
+    else {
+      // Ending up in a cycle (e.g. from `z = z + "foo";`) will mark the node as LR
+      nodeLRResults.set(node, true);
+      const wasLR = isLRWithoutResultCache(node);
+      nodeLRResults.set(node, wasLR);
+      return wasLR;
+    }
+  }
+
+  if (!isLR(callParam)) {
+    // TODO: report the node that failed the check, in order to get the right line number
+    const lineNumber = callParam.getSourceFile().getLineAndColumnAtPos(callParam.getStart()).line;
+    return ["sqlInjection", {lineNumber: lineNumber}];
+  }
+}
+
+const isSqlInjection: ErrorChecker = (node, fnDecl, _isLocal) => {
+  // TODO: support cases where it's just `client.raw` (without the `TransactionContext` part)
+  if (Node.isCallExpression(node)) {
+    const subexpr = node.getExpression();
+
+    // `ctxt.client.<callName>`
+    const identifiers = subexpr.getDescendantsOfKind(SyntaxKind.Identifier);
 
-// At the moment, this only performs analysis on expected-to-be-deterministic functions
-function analyzeFunction(fn: FunctionOrMethod) {
-  const body = fn.getBody();
+    // An injection in DBOS must match `ctxt.client.<something>`
+    if (identifiers.length !== 3) return;
```

Comment:
Does this work if I assign `ctxt.client` to another variable and invoke `raw`?
```
const client = ctxt.client;
client.raw(`sql stmt`);
```

Reply:
Not yet - I have a TODO for that